src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java

```diff
@@ -275,107 +422,164 @@ private String getClassName(JarEntry f) {
      * @param name name
      * @param predicate predicate
      * @return true if yes
+     *
+     * @see RuntimeEnvironment#getPluginConfiguration()
      */
     @SuppressWarnings("unchecked")
-    private boolean checkAll(HttpServletRequest request, String cache, String name,
-            Predicate<IAuthorizationPlugin> predicate) {
+    private boolean checkAll(HttpServletRequest request, String cache, Nameable entity,
+            Predicate<AuthorizationPluginWrapper> predicate) {
+        Boolean val;
         Map<String, Boolean> m = (Map<String, Boolean>) request.getAttribute(cache);
 
         if (m == null) {
             m = new TreeMap<>();
+        } else if ((val = m.get(entity.getName())) != null) {
+            // cache hit
+            return val;
         }
 
-        Boolean val = m.get(name);
+        boolean overallDecision = performCheck(entity, predicate);
 
-        if (val != null) {
-            return val;
-        }
+        m.put(entity.getName(), overallDecision);
+        request.setAttribute(cache, m);
+        return overallDecision;
+    }
 
-        for (IAuthorizationPlugin plugin : getPlugins()) {
+    private boolean performCheck(Nameable entity, Predicate<AuthorizationPluginWrapper> predicate) {
+        boolean overallDecision = true;
+        for (AuthorizationPluginWrapper plugin : getPlugins()) {
+            // run the plugin's test method
             try {
-                if (!predicate.test(plugin)) {
-                    m.put(name, false);
-                    request.setAttribute(cache, m);
-                    return false;
+                boolean pluginDecision = predicate.test(plugin);
+
+                if (!pluginDecision && plugin.isRequired()) {
+                    // required sets a failure but still invokes all other plugins
+                    overallDecision = false;
+                    continue;
+                } else if (!pluginDecision && plugin.isRequisite()) {
+                    // requisite sets a failure and immediately returns the failure
+                    overallDecision = false;
+                    break;
+                } else if (overallDecision && pluginDecision && plugin.isSufficient()) {
+                    // sufficient immediately returns the success
+                    overallDecision = true;
+                    break;
                 }
             } catch (Throwable ex) {
-                LOGGER.log(Level.SEVERE, "Plugin \"" + plugin.getClass().getName() + "\" has failed with exception:", ex);
-                removePlugin(plugin);
+                LOGGER.log(Level.WARNING,
+                        String.format("Plugin \"%s\" has failed the testing of \"%s\" with an exception.",
+                                plugin.getClassname(),
+                                entity.getName()),
+                        ex);
+
+                // set the return value to false for this faulty plugin
+                if (!plugin.isSufficient()) {
```

Comment:
I don't think so, as I read in the man:

>  A failure of a sufficient module is ignored and processing of the PAM module stack continues unaffected.

Sufficient plugin should only decide if the stack can be left immediately if it returns true. In all other cases (false, failure) it should not affect the stack processing

For the other types it could be that the stack will be left immediately when the plugin has a failure (or even returns false). The only argument to leave as it is is a timing attack but I don't think that it's that important here.